### PR TITLE
[FSDP] use all_gather for 10X OSD consolidation speedup

### PR DESCRIPTION
**Pull request by sshleifer**
TLDR: Using `all_gather` instead of `broadcast` for optimizer state consolidation appears to be a speed win without a memory cost.

### Approach:
- for tensor state in OSD we use `all_gather`
- for non-tensor metadata (`loss_scale`, `param_groups`, `num_padded`) we use `broadcast`.
- The same OSD (for 300M param model) takes 110 ms to consolidate vs 2300 ms, speedup usually around 10X.
- assume `recipient_rank=0`, since there are no other callers. 

### Evidence of Win:
- This appears not to use extra GPU RAM, and save a lot of time. For the unittests, the process goes from 2300 MS to consolidate a larger MOE to 110 Ms. 
- For a large fairseq 2.2B param model (1.2T config), it takes 80s instead of 800s, with less CUDA usage.
- fairseq resumption test (resume training from ckpt should result as same loss as training from scratch) passes.

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1402,20 +1405,58 @@ def _consolidate_optim_state_dict(
...
+                non_tensor_state.append(
+                    recursive_copy_to_device(send_data, non_blocking=False, device=torch.device("cpu"))
+                )
 
-        return all_states
+            self._print_r0(f"gathered non tensor state from rank {rank}")
+
+        return non_tensor_state, tensor_state
+
+    def _gather_optim_state(self, sd_state: Dict[int, Dict[str, Any]]) -> Dict[int, Dict[str, List]]:
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1433]: This is the new `_all_gather` logic

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1382,45 +1382,48 @@ def assert_state(self, state: Union[TrainingState, List[TrainingState]]) -> None
...
+
+                    if self.rank == 0:
+                        gathered_state[k][buffer_name] = [x.cpu() for x in chunks]
+                elif self.rank == 0:
+                    # Add non tensor state
+                    if buffer_name in gathered_state[k]:
+                        gathered_state[k][buffer_name].append(t)
+                    else:
+                        gathered_state[k][buffer_name] = [t]
+            self._print_r0(f"gathered tensor state for key {k} from all ranks")
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1423]: del

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1382,45 +1382,48 @@ def assert_state(self, state: Union[TrainingState, List[TrainingState]]) -> None
...
+            pad_info = broadcast_object(
+                pad_info, src_rank=rank, group=self.process_group, dist_device=self.compute_device
+            )
+            if self.rank == 0:
+                world_pad_info.append(pad_info)  # type: ignore
+        return world_pad_info
+
+    def _gather_optim_state(self, sd_state: Dict[int, Dict[str, Any]]) -> Dict[int, Dict[str, List]]:
+        """for each buffer in state[i], if the buffer is a tensor, collect it from the world. Else use rank 0's entry."""
+        self._print_r0(f"start: n state: {len(sd_state)}")
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1402]: del

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1430,24 +1433,37 @@ def gather_full_optim_state_dict(
...
-            a dict with two entries
+            a dict with four entries
                 * state - a dict holding gathered optimization state, 1 entry per unflat parameter
                 * param_groups - a dict containing the 1 parameter group
+                * param_id_map - global (unflat) to local (flat) id mapping
+                * uncollected_local_ids - keys in the state dict that were not broadcast
 
         """
+
+        self._tstart = time.time()
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1446]: del

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -1430,24 +1433,37 @@ def gather_full_optim_state_dict(
...
             return None
+        assert set(sd.keys()) == {"param_groups"}
         # Unify the shard states by concatenating tensors and unflattening params
         new_state_dict = ou.build_unflat_state_dict(
-            self._fsdp_instances, world_optim_states, self.uncollected_opt_state
+            self._fsdp_instances, pad_info, state, self.uncollected_opt_state, sd["param_groups"]
         )
+
         self.uncollected_opt_state = {}
+        self._print_r0("FSDP: done unflat")
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1466]: del

File: tests/nn/data_parallel/test_fsdp.py [hunk 6]
```diff
@@ -627,15 +627,18 @@ def __init__(self, group, wrapper_config, checkpoint_act=False, delay_before_fre
 
         # "expert" params are different on each rank
         torch.manual_seed(42 + group.rank())
-        d_expert = 16
-        expert = nn.Linear(d_expert, 4)
+        d_expert = 23
```
Comment (sshleifer) [inline tests/nn/data_parallel/test_fsdp.py:630]: make sure we unpad expert params correctly.

**Comment by sshleifer** [comment]
Planning to merge 10am PT tomorrow, barring further comments.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -1382,72 +1382,93 @@ def assert_state(self, state: Union[TrainingState, List[TrainingState]]) -> None
...
+                    gathered_state[k][buffer_name] = [t]
+
+        return gathered_state, singleton_state
+
+    def gather_full_optim_state_dict(self, optim: torch.optim.Optimizer, **ignored: Dict) -> Optional[Dict[str, Any]]:
         """Return the last known global optimizer state. The returned state is compatible with Pytorch, in that the
         sharded properties are not exposed. Multiple parameter groups are not yet supported.
 
         This should be called only on the root FSDP instance.
+        Nested FSDP instances are supported as long as they have the same world_size as the parent or world_size=1.
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1437]: nice

**Review by min-xu-ai — APPROVED** [review]
I don't have time to fully review this, but this looks good at a high level.

**Comment by min-xu-ai** [comment]
cc @QuentinDuval @prigoyal FYI

**Comment by min-xu-ai** [comment]
> 
> 
> Planning to merge 10am PT tomorrow, barring further comments.

sorry for the delay, I was out for most of the last week.

File: fairscale/nn/data_parallel/fsdp_optim_utils.py [hunk 8]
```diff
@@ -165,3 +162,11 @@ def build_unflat_state_dict(
         "param_groups": param_groups,
         "uncollected_local_ids": list(uncollected_opt_state.keys()),
     }
+
+
+def is_singleton_tensor(x: Any) -> bool:
+    """Is x a dimensionless tensor"""
+    if torch.is_tensor(x):
+        return not x.size()
```
Comment (myleott) [inline fairscale/nn/data_parallel/fsdp_optim_utils.py:170]:
```suggestion
        return x.dim() == 0
```

File: fairscale/nn/data_parallel/fsdp_optim_utils.py [hunk 9]
```diff
@@ -165,3 +162,11 @@ def build_unflat_state_dict(
...
+    else:
+        return False
```
Comment (myleott) [inline fairscale/nn/data_parallel/fsdp_optim_utils.py:172]:
```suggestion
    return torch.is_tensor(x) and x.dim() == 0
```

File: fairscale/nn/data_parallel/fsdp_optim_utils.py [hunk 10]
```diff
@@ -24,19 +25,26 @@ def flatten_optim_state_dict(sd: Dict) -> Dict:
...
-
     # Now combine all tensors in each buffer using torch.cat().
     for local_id, state in new_state.items():
         for buffer_name, tensors in state.items():
             new_state[local_id][buffer_name] = torch.cat(tensors)
         new_state[local_id].update(non_tensor_state)
+        new_state[local_id].update(singleton_state[local_id])
     new_sd = {"state": new_state, "param_groups": copy.deepcopy(sd["param_groups"])}
+    for k in sd.keys():  # if there are extra keys, like loss_scale, don't delete them
+        if k not in {"state", "param_groups", "uncollected_local_ids", "param_id_map"}:
```
Comment (myleott) [inline fairscale/nn/data_parallel/fsdp_optim_utils.py:46]: I'm slightly uneasy about this falling out of sync with line 160. Thoughts on some way to enforce parity between them?

**Review by myleott — APPROVED** [review]
LGTM!

File: fairscale/nn/data_parallel/fsdp_optim_utils.py (hunk 10 above)
Comment (sshleifer) [reply fairscale/nn/data_parallel/fsdp_optim_utils.py:46]: module level constant + comment + assert

```python
# These return keys are used by fairseq. To change, add @sshleifer as a reviewer.
UNFLAT_RETURN_KEYS = {"state", "param_groups", "uncollected_local_ids", "param_id_map"}
...
assert set(unflat_optim_state_dict.keys()) == UNFLAT_RETURN_KEYS 
```

File: fairscale/nn/data_parallel/fsdp_optim_utils.py (hunk 9 above)
Comment (sshleifer) [reply fairscale/nn/data_parallel/fsdp_optim_utils.py:172]: nice
